File: extensions/plugins/cn.vastplan.platform.artifacts.assessment.controller/controller/contribution.go
package controller

import (
	"context"
	"encoding/json"

	contractv1 "cdsoft.com.cn/VastPlan/core/shared/go/contract/v1"
	"cdsoft.com.cn/VastPlan/core/shared/go/extpoint"
	sdk "cdsoft.com.cn/VastPlan/extensions/sdk/go/plugin"
)

func (c *Controller) Contribution() sdk.Contribution {
	descriptor := []byte(`{"title":"插件制品持续复扫 Controller","subcommands":[{"name":"status","description":"读取低基数调度状态"}]}`)
	return sdk.Contribution{ExtensionPoint: extpoint.ToolPackage, ID: Capability, Descriptor: descriptor, Handlers: map[string]sdk.Handler{
		"status": func(context.Context, sdk.Host, *contractv1.CallContext, []byte) (*contractv1.CallResult, []byte, error) {
			raw, err := json.Marshal(c.Stats())
			return &contractv1.CallResult{Status: contractv1.CallResult_STATUS_OK}, raw, err
		},
	}}
}

File: extensions/sdk/node/configuration-controller/src/contribution.mjs
import { Contribution, callResult } from "@vastplan/backend-plugin";

import {
  CONFIGURATION_CONTROLLER_EXTENSION_POINT,
  CONFIGURATION_CONTROLLER_PROTOCOL,
  configurationControllerCapability,
  parseControllerRequest,
  validateObservation,
} from "./wire.mjs";

const operations = ["prepare", "commit", "abort", "status"];

export function configurationControllerContribution(pluginId, controller) {
  if (!controller || operations.some((operation) => typeof controller[operation] !== "function")) throw new Error("configuration.v1 controller 必须实现 prepare/commit/abort/status");
  return new Contribution({
    extensionPoint: CONFIGURATION_CONTROLLER_EXTENSION_POINT,
    id: configurationControllerCapability(pluginId),
    descriptor: { protocol: CONFIGURATION_CONTROLLER_PROTOCOL },
    handlers: Object.fromEntries(operations.map((operation) => [operation, handler(operation, controller)])),
  });
}

function handler(operation, controller) {
  return async (invocation, host, context, payload) => {
    const caller = context?.caller;
    if (!context?.tenant_id || !caller || ![3, "CALLER_KIND_PLUGIN"].includes(caller.kind) || caller.id !== "cn.vastplan.platform.configuration.plugin-settings") {
      return callResult.error("configuration.controller.permission_denied", "configuration.v1 只接受 plugin-settings 认证调用");
    }
    let request;
    try {
      request = parseControllerRequest(operation, payload);
    } catch (error) {
      return callResult.error("configuration.controller.invalid_request", error.message);
    }
    try {
      invocation.throwIfCancelled();
      const observation = validateObservation(await controller[operation](request, { invocation, host, context }));
      return callResult.ok(Buffer.from(JSON.stringify(observation)));
    } catch (error) {
      return callResult.error("configuration.controller.rejected", error.message);
    }
  };
}
